azure_img_utils/azure_image.py

```diff
@@ -149,33 +160,57 @@ def upload_image_blob(
         elif self.image_blob_exists(blob_name) and force_replace_image:
             self.delete_storage_blob(blob_name)
 
-        kwargs = {
-            'is_page_blob': is_page_blob,
-            'expand_image': expand_image
-        }
-
-        if max_workers:
-            kwargs['max_workers'] = max_workers
-
-        if max_retry_attempts:
-            kwargs['max_retry_attempts'] = max_retry_attempts
+        if max_retry_attempts < 0:
+            raise Exception(
+                f'max_retry_attempts parameter has to be >=0, '
+                f'{max_retry_attempts} provided.'
+            )
 
         try:
-            upload_azure_file(
-                blob_name,
-                self.container,
-                image_file,
+            blob_client = get_blob_client(
                 self.blob_service_client,
-                **kwargs
+                blob_name,
+                self.container
+            )
+
+            if is_page_blob:
+                blob_type = 'PageBlob'
+            else:
+                blob_type = 'BlockBlob'
+
+            system_image_file_type = FileType(image_file)
+            if system_image_file_type.is_xz() and expand_image:
+                open_image = lzma.LZMAFile
+            else:
+                open_image = open
+
+            msg = ''
+            attempts = 1 + max_retry_attempts
+            while attempts > 0:
```

Comment:
```
while max_retry_attempts != 0:
```
removes the correction for the index based counting and allows the user to not have to think about the details w.r.t. off by 1 for the `>` comparison.

Reply:
Agree, >= 0 is a better clause. It alone would handle negative values and prevent any infinite loop in those cases.

Reply:
But the original algorithm was correct? We exit when we decrement to zero and the loop runs a max number of times. No need to do a +1 before starting.

Reply:
The original algorithm uses 'max_retry_attempts' as 'max_attempts'  (if you set 1, you try just once).
May be it's my broken English, but with max_retry_attempts = 1 I'd expect 1 attempt and 1 retry, or with value 0 I'd expect the code to attempt once at least (with no retries).

Reply:
I agree `max_retry_attempts` states how often we should try again. So if `max_retry_attempts` is 1 then there would be a total of 2 attempts at the operation, 1 time for the attempt and 1 retry.

I am fine with `>=`

This is not an API change, it's a behavioral change. One could argue that based on naming of variables the original code had a bug, albeit the intend may have been different but did not match the variable naming.

Reply:
That's a bad variable name. The intention is that it is max_attempts. It seems more user friendly for the number of tries to be the number the user enters. Not that number plus one. My opinion would be rename variable to max_attempts and leave algorithm as it was in the function.

Reply:
OK, let me change it to max_attempts and update the cli accordingly.